packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -48,7 +48,10 @@ class ProtocolDiscovery {
   StreamSubscription<String> _deviceLogSubscription;
 
   /// The discovered service URI.
-  Future<Uri> get uri => _completer.future;
+  Future<Uri> get uri async {
+    final Uri rawUri = await _completer.future;
+    return await _forwardPort(rawUri);
```

Comment:
Specifically only try to `_forwardPort` when asked for, because if mDNS discovery succeeds, we'll never use this.